### PR TITLE
Annotator request access

### PR DESCRIPTION
**Pull request by Louise15926**

### CODE REVIEW

**Comment by ra1nty** [comment]
#47

**Review by ra1nty — CHANGES_REQUESTED** [review]
Seems you accidentally delete the content of `deploy_rsa.enc`, which is necessary for the CD.

File: instadam/project.py [hunk 1]
```diff
@@ -319,3 +320,74 @@ def update_user_permission(project_id):
...
+    project = Project.query.filter_by(id=project_id).first()
+    if project is None:
+        abort(404, 'Project with id=%s does not exist' % project_id)
+
+    current_user = get_jwt_identity()
+    user = User.query.filter_by(username=current_user).first()
+
+    map_code_to_message_type = {
+        'r_request': MessageTypeEnum.READ_ONLY_PERMISSION_REQUEST,
+        'rw_request': MessageTypeEnum.READ_WRITE_PERMISSION_REQUEST,
```
Comment (ra1nty) [inline instadam/project.py]: Should we consider changing the request parameter name to 'r' and 'rw' since 'request' is already included in the endpoint?

**Review by ra1nty — COMMENTED** [review]
Well-written test cases. Need some minor changes and fixes

**Comment by ra1nty** [comment]
#47

File: instadam/project.py (hunk 1 above)
Comment (Louise15926) [reply instadam/project.py]: Initially I was worried 'r' and 'rw' would be misleading if in the future we are not only dealing with project request-type of messages. But yeah for now leaving out 'request' might make more sense.

**Review by ra1nty — APPROVED** [review]

**Review by kuobao — COMMENTED** [review]
Checklist:
1. Usage of inheritance to avoid alternative classes with different interfaces
2. Sharing of functions to avoid code duplication
3. Well-formatted and easily understandable code
4. Correctly implement what is stated in the user stories
5. At least 90% of lines of the code should be tested
6. Error handlers for unexpected behaviors in tests
7. No unneeded procedures nor unreachable code

Comments:
Really clean and easily-understandable code overall. We did not find any code smells in your code and we think that you did a great job in refactoring. There are some small mistakes such as the comments on line 14 in message.py file which is probably mistyped. We think that it would be better if the MessageReceiverLink() function on line 49 of message.py could have some comments on what it does since it gave Charlie and me a hard time understanding the function. We think that your testing is pretty good even though we do not know the exact coverage. The function are pretty straight-forward and Charlie and I agreed that you do not need very complicated unit tests for this PR. However, we were not able to find corresponding user story (at least no name under your user story matches the PR title). In the end, we do not think there is much you can improve since the quality of your code is pretty good overall but adding more comments in each function could make your code more understandable for programmers who are not in your team.

**Comment by ra1nty** [comment]
> Checklist:
> 
> 1. Usage of inheritance to avoid alternative classes with different interfaces
> 2. Sharing of functions to avoid code duplication
> 3. Well-formatted and easily understandable code
> 4. Correctly implement what is stated in the user stories
> 5. At least 90% of lines of the code should be tested
> 6. Error handlers for unexpected behaviors in tests
> 7. No unneeded procedures nor unreachable code
> 
> Comments:
> Really clean and easily-understandable code overall. We did not find any code smells in your code and we think that you did a great job in refactoring. There are some small mistakes such as the comments on line 14 in message.py file which is probably mistyped. We think that it would be better if the MessageReceiverLink() function on line 49 of message.py could have some comments on what it does since it gave Charlie and me a hard time understanding the function. We think that your testing is pretty good even though we do not know the exact coverage. The function are pretty straight-forward and Charlie and I agreed that you do not need very complicated unit tests for this PR. However, we were not able to find corresponding user story (at least no name under your user story matches the PR title). In the end, we do not think there is much you can improve since the quality of your code is pretty good overall but adding more comments in each function could make your code more understandable for programmers who are not in your team.

Hi Kuobao & Charlie,

Thanks for your inspiring comments and reviews! We noticed our lack of documentation and will definitely make some improvement in the future. For the `MessageReceiverLink` Class, if you follow closely through the codebase or have some knowledge on SqlAlchemy, you will find it’s just a class for the association table which maintains the many-to-many relationship between the Message table and User table.

**Comment by Louise15926** [comment]
> Checklist:
> 
> 1. Usage of inheritance to avoid alternative classes with different interfaces
> 2. Sharing of functions to avoid code duplication
> 3. Well-formatted and easily understandable code
> 4. Correctly implement what is stated in the user stories
> 5. At least 90% of lines of the code should be tested
> 6. Error handlers for unexpected behaviors in tests
> 7. No unneeded procedures nor unreachable code
> 
> Comments:
> Really clean and easily-understandable code overall. We did not find any code smells in your code and we think that you did a great job in refactoring. There are some small mistakes such as the comments on line 14 in message.py file which is probably mistyped. We think that it would be better if the MessageReceiverLink() function on line 49 of message.py could have some comments on what it does since it gave Charlie and me a hard time understanding the function. We think that your testing is pretty good even though we do not know the exact coverage. The function are pretty straight-forward and Charlie and I agreed that you do not need very complicated unit tests for this PR. However, we were not able to find corresponding user story (at least no name under your user story matches the PR title). In the end, we do not think there is much you can improve since the quality of your code is pretty good overall but adding more comments in each function could make your code more understandable for programmers who are not in your team.

Thanks for your comment! I agree that we could do better on our comments especially within each function. Will definitely look into that. The reason why there is no corresponding user story is that this (annotator requesting access) is a newly added user story that was not included in early design. We will consider updating our user stories on the project page.

**Pull request merged**